File: tools/aivdm_translate.rb
#!/usr/bin/env ruby
require 'asciidoctor'

# What's ruby doing in a python library? Good question. The protocol information is documented
# in an asciidoc file, on the web at http://catb.org/gpsd/AIVDM.txt and included here. I can't
# find a good asciidoc library for python, so this ruby script is used to extract useful info
# from the protocol docs and turn it into JSON so the decoders can be built automatically.
#
# This is all a little ugly, but hopefully it will need to be used very rarely.

class ProtocolTable
  attr_reader :title, :headers, :rows

  def initialize(adt)
    @title = adt.title
    @headers = adt.rows.head[0].map { |c| c.text }
    @rows = adt.rows.body.map { |r| r.map { |c| c.text } }
    validate_headers
  end

  EXPECTED_HEADERS = %w(Field Len Description Member T Units)

  def validate_headers
    EXPECTED_HEADERS.zip(@headers).each do |expected, actual|
      raise "suspicious headers for #{@title}: #{expected} != #{actual}" unless expected==actual
    end
  end

  def fields_as_structure
    rows.map do |row|
      result = Hash[@headers.zip(row).map { |key, value| [key.downcase, value] }]
      s, e = row[0].split('-')
      result['start'] = s.to_i
      result['end'] = e.to_i
      result['len'] = result['len'].to_i
      result
    end
  end

  def as_structure
    {
        name: @title,
        messages: fields_as_structure
    }

  end
end


def find_doc_tables(block, default_title=nil)
  result = []
  if block.blocks?
    if block.title =~ /Type 7/
      puts("#{block.level} #{block.title}")
    end
    if block.render =~ /Type 7: Binary Acknowledge/
      puts ("-----foo-----")
      puts("#{block.level} #{block.render}")
      puts ("-----foo-----")
    end
    block.blocks.each do |child|
      if child.is_a? Asciidoctor::Table
        result << child
      end
      if child.blocks?
        result.concat(find_doc_tables(child))
      end
    end
  end
  result
end

doc = Asciidoctor.load(File.new('AIVDM.txt'))
doc_tables = find_doc_tables(doc)
doc_tables.each do |t|
  title = t.title
  if not title
    if t.level == 2
      title = t.parent.title
    else
      title == t.parent.parent.title
    end
  end
  # puts("#{t.level}\t#{title}\t#{t.title}\t#{t.parent.title}")
  # puts t.title
end
exit 0

sec = doc.blocks.last.blocks[18]

table = sec.blocks[1]

t = ProtocolTable.new(table)

require 'json'


results = {"message types": {"1": t.as_structure, "2": t.as_structure, "3": t.as_structure}}
puts(JSON.pretty_generate(results))

File: tools/aivdm_translate.py
#!/usr/bin/env python
import os
import re
import json
from collections import OrderedDict

from bs4 import BeautifulSoup


# Takes the AIVDM html file (downloaded from from http://catb.org/gpsd/AIVDM.html) and extracts JSON
# that is used by SimpleAIS to do protocol decoding. This only needs to be re-run when the source
# HTML changes or when expanding parser functionality. When using the library normally, all packet
# info is pulled from the AIVDM.json file.

soup = BeautifulSoup(open(os.path.join(os.path.dirname(__file__), 'AIVDM.html')), 'html.parser')


class Table:
    def __init__(self, soup_table, title_override=None):
        self.title = title_override or self.extract(soup_table, 'caption')
        self.headings = self.extract_all(soup_table, 'th')
        self.rows = [self.extract_all(row, 'td') for row in soup_table.find('tbody').find_all('tr')]

    def extract(self, soup_table, search):
        found = soup_table.find(search)
        if found:
            return found.text.strip()

    def extract_all(self, soup_table, search):
        found = soup_table.find_all(search)
        if found:
            return [f.text.strip() for f in found]

    def as_list_of_dicts(self):
        return [dict(zip(self.headings, row)) for row in self.rows]

    def __getitem__(self, item):
        return self.rows[item]


def fields_for_row(row):
    result = OrderedDict()
    m = re.search("(\d+)-(\d+)", row['Field'])
    if m:
        result['start'] = int(m.group(1))
        result['end'] = int(m.group(2))
    else:
        m = re.search("(\d+)", row['Field'])
        result['start'] = result['end'] = int(m.group(1))
    result['description'] = row['Description']
    if 'Member' in row:
        result['member'] = row['Member']
    elif 'Member/Type' in row:
        result['member'] = row['Member/Type']
    else:
        raise (ValueError("can't find member in row {}".format(row)))
    if 'T' in row:
        result['type'] = row['T']
    elif 'u' in row:
        result['type'] = row['u']
    else:
        raise (ValueError("can't find type in row {}".format(row)))
    return result


def message_info_for_table(table, name):
    result = OrderedDict()
    result['name'] = name
    result['fields'] = [fields_for_row(row) for row in table.as_list_of_dicts() if not row['Field'] == "?"]
    return result


message_types = Table(soup.find('h2', {'id': '_ais_payload_interpretation'}).find_next('table'))

messages = OrderedDict()

for h3 in soup.find_all('h3'):
    if re.search('Types 1, 2 and 3', h3.text):
        cnb_table = Table(h3.find_next('table'))
        messages["1"] = message_info_for_table(cnb_table, message_types[0][1])
        messages["2"] = message_info_for_table(cnb_table, message_types[1][1])
        messages["3"] = message_info_for_table(cnb_table, message_types[2][1])
    match = re.search('Type (\d+):?\s+(.*)', h3.text)
    if match:
        messages[match.group(1)] = message_info_for_table(Table(h3.find_next('table')), match.group(2))

print(json.dumps({'messages': messages}, indent=4, separators=(',', ': ')))
